dm/syncer/safe_mode.go

```diff
@@ -42,36 +56,68 @@ func (s *Syncer) enableSafeModeInitializationPhase(tctx *tcontext.Context) {
+	})
+	if initPhaseSeconds == "" {
+		duration = time.Second * time.Duration(2*s.cfg.CheckpointFlushInterval)
+	} else {
+		duration, err = time.ParseDuration(initPhaseSeconds)
+		if err != nil {
+			s.tctx.L().Error("enable safe-mode failed due to duration parse failed", zap.String("duration", initPhaseSeconds))
+			return
+		}
+	}
+	exitPoint := s.checkpoint.SafeModeExitPoint()
+	failpoint.Inject("SafeModeDurationIsZero", func() {
+		*s.beginLocation = binlog.MustZeroLocation(mysql.MySQLFlavor)
+	})
+	if exitPoint != nil {
 		//nolint:errcheck
 		s.safeMode.Add(tctx, 1) // enable and will revert after pass SafeModeExitLoc
-		s.tctx.L().Info("enable safe-mode for safe mode exit point, will exit at", zap.Stringer("location", *s.checkpoint.SafeModeExitPoint()))
+		s.tctx.L().Info("enable safe-mode for safe mode exit point, will exit at", zap.Stringer("location", *exitPoint))
 	} else {
-		//nolint:errcheck
-		s.safeMode.Add(tctx, 1) // enable and will revert after 2 * CheckpointFlushInterval
-		go func() {
-			defer func() {
-				err := s.safeMode.Add(tctx, -1)
-				if err != nil {
-					// send error to the fatal chan to interrupt the process
-					s.runFatalChan <- unit.NewProcessError(err)
-				}
-				if !s.safeMode.Enable() {
-					s.tctx.L().Info("disable safe-mode after task initialization finished")
+		s.tctx.L().Info("enable safe-mode because of task initialization", zap.Duration("duration", duration))
+
+		if int64(duration) > 0 {
+			//nolint:errcheck
+			s.safeMode.Add(tctx, 1) // enable and will revert after 2 * CheckpointFlushInterval
+			go func() {
+				defer func() {
+					err = s.safeMode.Add(tctx, -1)
+					if err == nil && !s.safeMode.Enable() {
+						s.tctx.L().Info("disable safe-mode after task initialization finished")
+					}
+				}()
+
+				select {
+				case <-tctx.Context().Done():
+				case <-time.After(duration):
 				}
 			}()
-			initPhaseSeconds := s.cfg.CheckpointFlushInterval * 2
-			failpoint.Inject("SafeModeInitPhaseSeconds", func(val failpoint.Value) {
-				seconds, _ := val.(int)
-				initPhaseSeconds = seconds
-				s.tctx.L().Info("set initPhaseSeconds", zap.String("failpoint", "SafeModeInitPhaseSeconds"), zap.Int("value", seconds))
-			})
-			s.tctx.L().Info("enable safe-mode because of task initialization", zap.Int("duration in seconds", initPhaseSeconds))
-			select {
-			case <-tctx.Context().Done():
-			case <-time.After(time.Duration(initPhaseSeconds) * time.Second):
-			}
-		}()
+		}
+	}
+
+	// when exitPoint equal beginLocation, it means safe mode doesn't enable
+	// because safe mode will disable quickly in the syncer main logic
+	if exitPoint != nil && binlog.CompareLocation(*exitPoint, *s.beginLocation, s.cfg.EnableGTID) == 0 {
+		s.tctx.L().Info("safe-mode: exitPoint equal to beginLocation")
+		return
+	}
+	if s.safeMode.Enable() && int64(duration) == 0 {
+		fresh, err2 := s.IsFreshTask(tctx.Ctx)
```

Comment:
Why fresh should be treated differently

Reply:
> Determine if it is the first time to start the task this machine.

Reply:
> Determine if it is the first time to start the task this machine.
> 
> If it is, `duration = 0` doesn't matter.

It's the first time to start sync, why it doesn't matter? If user use --consistency none for dumpling, I think we should deal with it aslo